src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -0,0 +1,262 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal static class PipeExtensions
+    {
+        private const int MessageDelimiterSize = 4; // how many bytes it takes to encode "Message-Length"
+        private const int HeaderSize = MessageDelimiterSize + 1; // message length + compression flag
+
+        public static Task WriteMessageAsync(this PipeWriter pipeWriter, byte[] messageData, bool flush = false)
+        {
+            WriteHeader(pipeWriter, messageData.Length);
+            pipeWriter.Write(messageData);
```

Comment:
We've evolved this logic a bit. See https://github.com/aspnet/AspNetCore/pull/7227/files#diff-d2000ab873598422d9bacae788b02a8bR56